### DISCUSSION TITLE
Glowing effect in IM7

### DISCUSSION THREAD

**Discussion by Rchardon** [ImageMagick/ImageMagick#6799] [Help] [answered]
### ImageMagick version

7.1.1-20

### Operating system, version and so on

Windows 10

### Description

i'm trying to fix my "add a glowing effect to my pngs" script. It was working in IM6 but not in IM7.

The command in IM6 was the following : convert my_file.png ( +clone -channel A -blur 0x2.5 -level 0,80% +channel +level-colors white ) -compose DstOver -composite png32:my_new_file.png
This would transform this 
![Sawblade](https://github.com/ImageMagick/ImageMagick/assets/33094562/90c7277d-3e36-4d64-847e-822ff17512aa)
 to this 
![Sawblade](https://github.com/ImageMagick/ImageMagick/assets/33094562/a99e14e4-d822-4816-b178-186354aa3518)

But with IM7, the command is the same except i replaced convert with magick and the command transforms this 
![Holy Poop](https://github.com/ImageMagick/ImageMagick/assets/33094562/16c810d0-ce22-4cbc-9588-8f57a81e4e59)
to this
![Holy Poop](https://github.com/ImageMagick/ImageMagick/assets/33094562/1e46829f-f7f4-4ff0-ab6f-284429fb8446)
There's now a white background instead of the glowing effect.

The guy who wrote the script originally isn't me so i'm a bit lost with IM commands

**Comment by fmw42** [comment]
There are differences in Imagemagick 7 from Imagemagick 6.  In Imagemagick 7, you need to enable transparency for all channels.

So add -alpha on -channel rgba to the command after reading the input.

Unix Syntax:

    magick my_file.png -alpha on -channel rgba \( +clone -channel A -blur 0x2.5 -level 0,80% +channel +level-colors white \) -compose DstOver -composite png32:my_new_file2.png

Windows Syntax:

    magick my_file.png -alpha on -channel rgba ( +clone -channel A -blur 0x2.5 -level 0,80% +channel +level-colors white ) -compose DstOver -composite png32:my_new_file2.png 

Result:

[![enter image description here][1]][1]


  [1]: https://i.stack.imgur.com/NHyEq.png

**Comment by Rchardon** [comment]
I copied the Windows command and i got this 
![Sawblade](https://github.com/ImageMagick/ImageMagick/assets/33094562/e473ac3b-5665-495b-89b0-f97040095fcc)

I'll paste the command from console : 
magick ".\collectibles\custom\Sawblade.png" -alpha on -channel rgba ( +clone -channel A -blur 0x2.5 -level 0,80% +channel +level-colors white ) -compose DstOver -composite png32:".\collectibles\custom\glow\Sawblade.png"

Do i need something else ? I don't get why i don't have the same results than you
Running my python script or pasting the command directly in the console doesn't change anything

**Reply by fmw42** [reply]
Ii only tested on the one image and got a similar result.  If you are putting that command in a .bat file, you need to double your % to  %%. Other than that, I do not know what would be wrong. It seemed to work fine in Unix for the one image.

**Reply by Rchardon** [reply]
I'm not using a bat. I run my python script directly via VSCode console and i also ran the command directly in a command prompt with %% and none worked. For now it seems to be a Windows problem

**Reply by fmw42** [reply]
In command prompt use % not %%.  I do not understand what you are doing with Python and Imagemagick. But the first thing is to get it working in a command prompt window. The command that I gave should have worked. It works for me on my Mac in the Unix syntax.  The only difference is that in Windows you do not escape the parentheses with backslashes.  Be sure to keep spaces between the parentheses and any text on either side.  Test with the first image and see if you can duplicated my result.  Note, that since it has transparency, you will need to view results outside a browser in some image editor/display tool.  Be sure you are using the very same input image that I did.

**Reply by Rchardon** [reply]
![Capture](https://github.com/ImageMagick/ImageMagick/assets/33094562/39785b53-5ed6-474b-89e5-9d5bd050a40d)
I'm just running the command in my script, nothing elaborated. The command doesn't work in a command prompt window. And i literally copied the Windows command you gave me and using the same image as you did

**Reply by Rchardon** [reply]
If you have an other command, an other way to do this i'll gladly take it

**Reply by fmw42** [reply]
Run the command without all the other code you show.. I do not know Windows, so do not know what language or tool that is. Put in the actual files, not variables to the command and run it directly in a command window.

**Reply by fmw42** [reply]
Why is the % and the following +c (on channel) in blue.  That might indicate some issue with % that might need to be escaped in some way.

**Reply by Rchardon** [reply]
When i say it doesn't work, it gives me the image above, so of course i put the real file in the command... As for the characters in blue i don't know, probably vscode related. Doesn't matter since the command doesn't work in a command prompt, don't focus on the code

**Comment by Rchardon** [comment]
If you have a better way to do a glowing effect i'll gladly take it. I tested many things without success, i'm out of ideas

It still works on IM6 but since i was using convert, Windows was complaining and adding convert in my Path variable didn't change anything, maybe i did that wrong. So i was forced to use the full path to call convert in my script which i didn't like since the script is in a public Github repo

**Reply by fmw42** [reply]
If that does not help, then wait for a Windows Imagemagick user to help further.

**Reply by Rchardon** [reply]
Ok, thanks for the help tho

**Reply by fmw42** [reply]
If that is viscose, then it might need to have the % escaped or the argument put in quotes. You can try the latter. I asked if you would run my command in your command window without vs code.  What was the result?

**Reply by fmw42** [reply]
Try 

`magick my_file.png -alpha on -channel rgba ( +clone -channel A -blur 0x2.5 -level "0,80%" +channel +level-colors white ) -compose DstOver -composite png32:my_new_file2.png` 

with quotes around the -level argument

**Comment by snibgo** [comment, marked as answer]
I didn't know what "`+level-colors white`", with only one colour specified, should do. So I read the documentation https://imagemagick.org/script/command-line-options.php#level-colors

> By supplying a single color with a comma separator either before or after that color, will just replace the respective 'black' or 'white' point respectively. But if no comma separator is provided, the given color is used for both the black and white color points, making the operator either threshold the images around that color (- form) or set all colors to that color (+ form). 

Hmm. I think this means that "`+level-colors white`" should do the same as "`+level-colors white,white`", which should set all colours to white, while retaining transparency. I have a vague memory of a change in the default, but can't find it in ChangeLog.md.

Anyhow, it seems a weird method of changing colours to white. I generally use "`-fill White -colorize 100`". So I suggest:
```
magick cog.png ( +clone -channel A -blur 0x2.5 -level 0,80%% +channel -fill White -colorize 100 ) -compose DstOver -composite png32:y.png
```
... where `cog.png` is the OP input image.

This seems to give the expected result when used in v6 (with `convert`) or v7.

EDIT: I tested it in a BAT file, so the % is doubled to %%. If non-BAT, use single %.

**Reply by Rchardon** [reply]
oh my god, my saviour ! Seems like +level-colors changed in IM7 then. Many thanks :)

**Comment by snibgo** [comment]
The difference is that in v6, the default `-channel` setting was `RGB`, so excluded the alpha channel. In v7, the default is `RGBA`, which includes the alpha channel. So in v6, `+level-colors white` by default retained the alpha values, but in v7, it changes alpha to opaque.

We can precede `+level-colors white` with an explicit `-channel` setting, like this:
```
magick cog.png  ( +clone -channel A -blur 0x2.5 -level 0,80% -channel RGB +level-colors white +channel ) -compose DstOver -composite x.png
```

**Reply by Rchardon** [reply]
The thing that i don't understand is that my command worked for the Unix user in IM7

**Comment by snibgo** [comment]
> The thing that i don't understand is that my command worked for the Unix user in IM7

I don't know. Perhaps the Unix user had an older v7 version, and the change to the default RGBA occurred after that. Not all the v7 modifications occurred at the same time.

**Reply by fmw42** [reply]
I was the user and it worked for me on IM 7.1.1.20 with my change enabling transparency as in my command above.

**Reply by snibgo** [reply]
Okay. For me, IM v7.1.1-20 on Windows 11, your command doesn't make the required result.
```
magick cog.png  -alpha on -channel rgba ( +clone -channel A -blur 0x2.5 -level 0,80% +channel +level-colors white ) -compose DstOver -composite png32:m.png
```
m.png is:

![m](https://github.com/ImageMagick/ImageMagick/assets/33812368/08ed541e-8c21-46d4-b395-c5136e3f035f)

The input file has an alpha channel, so `-alpha on` should have no effect.

**Reply by fmw42** [reply]
Your command should work better. I tried mine again today on my Mac and it did not work. Not sure why it worked yesterday. Very strange. But your command makes more sense.
